File: src/main/java/io/alpere/common/crudfop/model/BaseEntity.java
package io.alpere.common.crudfop.model;

import lombok.*;
import org.springframework.data.annotation.CreatedBy;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedBy;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import javax.persistence.*;
import java.time.ZonedDateTime;
import java.util.UUID;

@Getter
@Setter
@ToString(of = "id")
@EqualsAndHashCode(of = {"id"})
@NoArgsConstructor
@MappedSuperclass
@EntityListeners(AuditingEntityListener.class)
public abstract class BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(name = "id", nullable = false, updatable = false)
    protected UUID id;

    @CreatedBy
    @Column(name = "created_by", nullable = false, updatable = false)
    protected UUID createdBy;

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    protected ZonedDateTime createdAt;

    @LastModifiedBy
    @Column(name = "updated_by", nullable = false)
    protected UUID updatedBy;

    @Version
    @LastModifiedDate
    @Column(name = "updated_at", nullable = false)
    private ZonedDateTime updatedAt;


    @Column(name = "deleted_by")
    protected UUID deletedBy;

    @Column(name = "deleted_at")
    private ZonedDateTime deletedAt;

}

File: src/main/kotlin/io/alpere/crud/model/BaseEntity.kt
package io.alpere.crud.model

import org.springframework.data.util.ProxyUtils
import java.util.*
import javax.persistence.*

@MappedSuperclass
abstract class BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", nullable = false, updatable = false)
    var id: UUID? = null

    override fun equals(other: Any?): Boolean {
        other ?: return false

        if (this === other) return true

        if (javaClass != ProxyUtils.getUserClass(other)) return false

        other as BaseEntity

        return this.id != null && this.id == other.id
    }

    override fun hashCode() = 19

    override fun toString(): String {
        return "${this.javaClass.simpleName}(id=$id)"
    }
}